images/cfw/opt/x1plusd.py

```diff
@@ -1,116 +1,282 @@
 logger.setLevel(logging.DEBUG)
 
-slh = logging.handlers.SysLogHandler(address = '/dev/log')
+slh = logging.handlers.SysLogHandler(address="/dev/log")
 slh.setLevel(logging.INFO)
-slh.setFormatter(logging.Formatter('%(name)s: %(message)s'))
+slh.setFormatter(logging.Formatter("%(name)s: %(message)s"))
 logger.addHandler(slh)
 
 ch = logging.StreamHandler()
 ch.setLevel(logging.DEBUG)
-ch.setFormatter(logging.Formatter('[%(asctime)s] %(name)s: %(levelname)s: %(message)s'))
+ch.setFormatter(logging.Formatter("[%(asctime)s] %(name)s: %(levelname)s: %(message)s"))
 logger.addHandler(ch)
 
+
+class OTACheckService:
+    def __init__(self, router):
+        self.router = router
+        self.ota_url = "https://ota.x1plus.net/stable/ota.json"
+        self.ota_available = False
+        self.last_check_timestamp = None
+        self.last_check_response = None
+        self.last_check_error = False
+        self.ota_downloaded = False
+
+        try:
+            with open("/opt/info.json", "r") as fh:
+                self.build_info = json.load(fh)
+        except FileNotFoundError:
+            logger.warning(
+                "OTA engine did not find /opt/info.json, Setting mock values so we get an OTA to recover!"
+            )
+            self.build_info = {
+                "cfwVersion": "0.1",
+                "date": "2024-04-17",
+                "buildTimestamp": 1713397465.0,
+            }
+
+    async def task(self):
+        # Catch our OTA calls
+        OTAMatch = MatchRule(
+            interface="x1plus.ota", path="/x1plus/ota", type="method_call"
+        )
+        await Proxy(message_bus, self.router).AddMatch(OTAMatch)
+        with self.router.filter(OTAMatch, bufsize=0) as queue:
+            while True:
+                msg = await queue.get()
+                method = msg.header.fields[HeaderFields.member]
+
+                if msg.header.fields[HeaderFields.signature] != "s":
+                    await self.router.send(
+                        new_error(msg, "x1plus.x1plusd.Error.BadSignature")
+                    )
+                    continue
+
+                arg = msg.body[0]
+
+                rv = None
+                try:
+                    if method == "ota":
+                        rv = await self._update_check(json.loads(arg))
+
+                except Exception as e:
+                    logger.error(f"{method}({arg}) -> exception {e}")
+                    await self.router.send(
+                        new_error(msg, "x1plus.x1plusd.Error.InternalError")
+                    )
+                    continue
+
+                if not rv:
+                    logger.warning(f"{method} -> NoMethod")
+                    await self.router.send(
+                        new_error(msg, "x1plus.x1plusd.Error.NoMethod")
+                    )
+                    continue
+
+                logger.debug(f"{method}({arg}) -> {rv}")
+                await self.router.send(new_method_return(msg, "s", (rv,)))
```

Comment:
It would be nice to refactor the 'Dbus RPC catcher', since this otherwise repeats the thing below, and it would be nice not to have that coded xeroxed.

Reply:
Maybe we define a parent class that defines this and we inherit it for settings and OTA? We can think on it as I'm going to strip OTA from this PR and move it to its own draft PR after the settings logic is merged.

Reply:
That’s kind of my thought, yeah.